cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -151,30 +151,27 @@ class SeqVisitor {
     memset(nesting_histogram_, 0, MAX_NESTING_LEVELS * sizeof(int));
   }
 
+  // co-recursive with VisitElem
   Status Visit(PyObject* obj, int level = 0) {
-    Py_ssize_t size = PySequence_Size(obj);
-
     if (level > max_nesting_level_) { max_nesting_level_ = level; }
 
-    for (int64_t i = 0; i < size; ++i) {
-      // TODO(wesm): Error checking?
-      // TODO(wesm): Specialize for PyList_GET_ITEM?
-      OwnedRef item_ref(PySequence_GetItem(obj, i));
-      PyObject* item = item_ref.obj();
-
-      if (PyList_Check(item)) {
-        RETURN_NOT_OK(Visit(item, level + 1));
-      } else if (PyDict_Check(item)) {
-        return Status::NotImplemented("No type inference for dicts");
-      } else {
-        // We permit nulls at any level of nesting
-        if (item == Py_None) {
-          // TODO
-        } else {
-          ++nesting_histogram_[level];
-          scalars_.Visit(item);
-        }
+    // Loop through either a sequence or an iterator.
+    // VisitElemt takes ownership of the object we are visiting.
+    if (PySequence_Check(obj)) {
+      Py_ssize_t size = PySequence_Size(obj);
+      for (int64_t i = 0; i < size; ++i) {
+	// TODO(wesm): Specialize for PyList_GET_ITEM?
+	RETURN_NOT_OK(VisitElem(PySequence_GetItem(obj, i), level));
       }
+    } else if (PyObject_HasAttrString(obj, "__iter__")) {
+      PyObject* iter = PyObject_GetIter(obj);
+      PyObject* item;
+      while ((item = PyIter_Next(iter))) {
+	RETURN_NOT_OK(VisitElem(item, level));
+      }
+      Py_DECREF(iter);
```

Comment:
Put `iter` in an `OwnedRef` -- this fails before iteration finishes, this will leak memory